torch/_functorch/aot_autograd.py

```diff
@@ -719,223 +749,420 @@ def inner(*flat_args):
+                ]
             ]
-        if any(
-            meta.input_info[i].mutates_data
-            for i in outer_aliased_indices_of_current_base_arg
-        ):
-            # Make sure the primal we pass to autograd.grad()
-            # sees the tensor before the mutation
-            return t.clone()
-        if any(
-            meta.input_info[i].mutates_metadata and not meta.input_info[i].mutates_data
-            for i in outer_aliased_indices_of_current_base_arg
-        ):
-            # Make sure the primal we pass to autograd.grad()
-            # sees the tensor before the metadata mutation
-            return t.view(t.shape)
+        )
+        self.num_outputs_aliased_to_intermediates = len(
+            [
+                x
+                for x in self.fw_metadata.output_info
+                if x.output_type in [
+                    OutputType.alias_of_intermediate,
+                    OutputType.alias_of_intermediate_save_as_output,
+                    OutputType.alias_of_intermediate_base_is_user_output,
+                ]
+            ]
+        )
+        self.num_outputs_aliased = (
+            self.num_outputs_aliased_to_inputs + self.num_outputs_aliased_to_intermediates
+        )
+        self.num_mutated_data_inputs = len(
+            [x for x in self.fw_metadata.input_info if x.mutates_data]
+        )
+        self.num_mutated_metadata_inputs = len(
+            [
+                x
+                for x in self.fw_metadata.input_info
+                if x.mutates_metadata
+            ]
+        )
+        self.num_mutated_metadata_only_inputs = len(
+            [
+                x
+                for x in self.fw_metadata.input_info
+                if not x.mutates_data and x.mutates_metadata
+            ]
+        )
+        self.num_mutated_inputs = self.num_mutated_data_inputs + self.num_mutated_metadata_only_inputs
+
+# This function takes in a tensor t, and returns one of t, t.view(), or t.clone().
+# When tracing the joint forward + backward, for any inputs in the graph that are mutated,
+# we need to clone them first (and similarly for metadata-only mutations, we need to view them first).
+# The idea is that when we trace the backward, we need to pass in the *original* primals
+# to autograd.grad(), before they were mutated.
+# Note: when we have synthetic base inputs, we need to clone them *before* creating views off of them.
+# This means that "idx" here represents the index of the (potentially) synthetic base.
+# What we need to do is:
+# (1) map the current (post-synthetic-base calling convention) input argument index
+#     to int index pre-synthetic-base-calling-convention.
+# (2) There could be multiple, if this index corresponds to a synthetic base
+#     that has multiple input aliases.
+# (3) If any of those corresponding inputs get metadata mutations, then we clone the base.
+def maybe_to_fresh_input(idx, t, meta):
+    if not isinstance(t, Tensor):
```

Comment:
When is this not a tensor?

Reply:
Ignoring dynamo for a second - if you directly pass a non-tensor input as an example input into `aot_module_simplified` or `aot_function`, it'll get plumbed down to here (AOTAutograd flattens pytrees, but it doesn't do anything to filter out non-tensor inputs). E.g. if you pass an `int` int (test [here](https://github.com/pytorch/pytorch/blob/a27bd42bb9ad39504fdd94ad38a5ad0346f1758b/test/functorch/test_aotdispatch.py#L334)).

Dynamo won't pass in int inputs, but some people are directly using the AOTAutograd API's, so I guess it's safer to be resilient to them.

SymInt is probably a better example (you can reasonably have a graph where its inputs are SymInts? Ed/Horace had some discussion around whether or not to establish some invariants around all SymInt inputs getting promoted to 0-dim tensors, although that's not the case today).